File: development/C++/src/server/Ae/ServerAeAddressSpaceElement.cpp
#include "OSCompat.h"
#include "ServerAeAddressSpaceElement.h"

#include "../ServerApplication.h"
#include "OTServer.h"

using namespace SoftingOPCToolboxServer;

//-----------------------------------------------------------------------------
//	AeAddressSpaceElement    - default constructor
//
AeAddressSpaceElement::AeAddressSpaceElement(void)
{
}   //  end ctr


//-----------------------------------------------------------------------------
//	AeAddressSpaceElement    -  constructor
//
AeAddressSpaceElement::AeAddressSpaceElement(
	tstring aName,
	unsigned long anUserData,
	unsigned long anObjectHandle,
	unsigned long aParentHandle) :
	AddressSpaceElement(aName, anUserData, anObjectHandle, aParentHandle)
{
}   //  end ctr


//-----------------------------------------------------------------------------
//	AeAddressSpaceElement    - destructor
//
AeAddressSpaceElement::~AeAddressSpaceElement(void)
{
}   //  end destructor


//-----------------------------------------------------------------------------
//	Adds a child to the current Address space element
//
BOOL AeAddressSpaceElement::addChild(AddressSpaceElement* aChild)
{
	if (!m_hasChildren)
	{
		Application::Instance()->trace(EnumTraceLevel_WRN, EnumTraceGroup_OPCSERVER,
									   _T("AeAddressSpaceElement::addChild"),
									   _T("this element does not allow children"));
		return FALSE;
	}   //  end if

	AeAddressSpaceElement* pChild = (AeAddressSpaceElement*)aChild;

	if (pChild == NULL)
	{
		Application::Instance()->trace(EnumTraceLevel_WRN, EnumTraceGroup_OPCSERVER,
									   _T("AeAddressSpaceElement::addChild"), _T("Invalid child provided"));
		return FALSE;
	}   //  end if

	AeAddressSpaceRoot* aeRoot = Application::Instance()->getAeAddressSpaceRoot();

	//  only add the child if no handle assigned
	if (pChild->getObjectHandle() == 0)
	{
		long result = E_FAIL;
		OTSAddressSpaceElementData data;
		data.m_name = (OTChar*)OTAllocateMemory((unsigned long)(pChild->getName().size() + 1) * sizeof(OTChar));

		if (data.m_name != NULL)
		{
#ifdef TBC_OS_WINDOWS
			_tcscpy(data.m_name, pChild->getName().c_str());
#endif
#ifdef TBC_OS_LINUX
			strcpy(data.m_name, pChild->getName().c_str());
#endif
		}

		tstring itemId = m_name + Application::Instance()->getAddressSpaceDelimiter() + pChild->getName();
		data.m_itemID = (OTChar*)OTAllocateMemory((unsigned long)(itemId.size() + 1) * sizeof(OTChar));

		if (data.m_itemID != NULL)
		{
#ifdef TBC_OS_WINDOWS
			_tcscpy(data.m_itemID, itemId.c_str());
#endif
#ifdef TBC_OS_LINUX
			strcpy(data.m_itemID, itemId.c_str());
#endif
		}

		data.m_hasChildren = (unsigned char)pChild->getHasChildren();
		data.m_isBrowsable = (unsigned char)pChild->getIsBrowsable();
		data.m_elementType = (unsigned char)EnumAddressSpaceElementType_AE;
		data.m_userData = pChild->getUserData();

		if (aeRoot->addElementToArray(pChild))
		{
			unsigned long childObjectHandle = 0;
			result = OTSAddAddressSpaceElement(m_objectHandle, &data, &childObjectHandle);
			pChild->setObjectHandle(childObjectHandle);
		}   //  end if

		//  release the memory allocated
		if (data.m_name != NULL)
		{
			OTFreeMemory(data.m_name);
			data.m_name = NULL;
		}   //  end if

		if (data.m_itemID != NULL)
		{
			OTFreeMemory(data.m_itemID);
			data.m_itemID = NULL;
		}   //  end if

		if (!SUCCEEDED(result))
		{
			result = aeRoot->removeElementFromArray(aChild);

			if (!SUCCEEDED(result))
			{
				//  Odd situation, just log an error
				Application::Instance()->trace(EnumTraceLevel_ERR, EnumTraceGroup_OPCSERVER,
											   _T("AeAddressSpaceElement::addChild"), _T("RemoveElementFromArray has failed"));
			}   //  end if

			delete aChild;
			return FALSE;
		}   //  end if ... else
	}
	else
	{
		aeRoot->addElementToArray(pChild);
	}   //  end if ... else

	return AddressSpaceElement::addChild(aChild);
}   //  end addChild


//-----------------------------------------------------------------------------
// Removes Child and if there, its children too
//
// returns:
// true     - Child removed
// false    - Child was not removed
BOOL AeAddressSpaceElement::removeChild(AddressSpaceElement* aChild)
{
	if (!m_hasChildren)
	{
		Application::Instance()->trace(EnumTraceLevel_WRN, EnumTraceGroup_OPCSERVER,
									   _T("AeAddressSpaceElement::removeChild"),
									   _T("this element does not allow children"));
		return FALSE;
	}   //  end if

	AeAddressSpaceElement* pAeChild = (AeAddressSpaceElement*)(aChild);

	if (pAeChild == NULL)
	{
		Application::Instance()->trace(EnumTraceLevel_WRN, EnumTraceGroup_OPCSERVER,
									   _T("AeAddressSpaceElement::removeChild"), _T("Invalid child provided"));
		return FALSE;
	}   //  end if

	if (pAeChild->getHasChildren())
	{
		//  Remove the clildren of the child
		std::vector<AddressSpaceElement*> childrenList = pAeChild->getChildren();
		std::vector<AddressSpaceElement*>::iterator pos;

		for (pos = childrenList.begin(); pos != childrenList.end(); pos++)
		{
			AddressSpaceElement* element = *pos;

			if (element != NULL)
			{
				pAeChild->removeChild(element);
			}   //  end if
		}   //  end for
	}   //  end if

	long result = S_OK;
	AddressSpaceElement::removeChild(aChild);
	unsigned long childHandle = pAeChild->getObjectHandle();
	AeAddressSpaceRoot* aeRoot = Application::Instance()->getAeAddressSpaceRoot();
	result = aeRoot->removeElementFromArray(pAeChild);

	if (!SUCCEEDED(result))
	{
		Application::Instance()->trace(EnumTraceLevel_WRN, EnumTraceGroup_OPCSERVER,
									   _T("AeAddressSpaceElement::removeChild"), _T("removeElementFromArray has failed"));
	}   //  end if

	delete pAeChild;
	pAeChild = NULL;

	if (childHandle != 0)
	{
		result = OTSRemoveAddressSpaceElement(childHandle);
	}   //  end if

	if (!SUCCEEDED(result))
	{
		Application::Instance()->trace(EnumTraceLevel_WRN, EnumTraceGroup_OPCSERVER, _T("AeAddressSpaceElement::removeChild"),
									   _T("OTSRemoveAddressSpaceElement has failed"));
		return FALSE;
	}   //  end if

	return TRUE;
}   //  end removeChild


//-----------------------------------------------------------------------------
//	getParent
//
//		-   Returns the element`s parent
//
AddressSpaceElement* AeAddressSpaceElement::getParent(void)
{
	if (m_objectHandle == 0)
	{
		return NULL;
	}   //  end if

	OTObjectData parent;

	if (S_OK != OTSGetParent(m_objectHandle, &parent))
	{
		return NULL;
	}//end if

	AeAddressSpaceRoot* aeRoot = Application::Instance()->getAeAddressSpaceRoot();

	if (aeRoot == NULL)
	{
		return NULL;
	}   //  end if

	return aeRoot->getElementFromArray(parent.m_userData);
}   //  end getParent


//-----------------------------------------------------------------------------
//	getChildren
//
//	-    Returns a list of children.
//		this list is empty in case of an error or if no children are there
//		to be delivered
//
std::vector<AddressSpaceElement*> AeAddressSpaceElement::getChildren(void)
{
	std::vector<AddressSpaceElement*> children;
	OTSObjectData* ptrData = NULL;
	long count = 0;

	if (SUCCEEDED(OTSGetChildren(m_objectHandle, EnumAddressSpaceElementType_AE, &count, &ptrData)))
	{
		AeAddressSpaceRoot* aeRoot = Application::Instance()->getAeAddressSpaceRoot();
		AddressSpaceElement* element = NULL;

		if (aeRoot != NULL)
		{
			aeRoot->m_elementsJanitor.lock();

			for (int i = 0; i < count; i++)
			{
				element = aeRoot->getElementFromArray(ptrData[i].m_userData);
				if (element != NULL)
				{
					children.push_back(element);
				}   //  end if
			}   //  end for

			aeRoot->m_elementsJanitor.unlock();
		}

		OTFreeMemory(ptrData);
	}   //  end if

	return children;
}   //  end getChildren


//-----------------------------------------------------------------------------
//	queryConditions
//
//	-    used to return the list of connditions names of the current element for
//		for a specific path
//
long AeAddressSpaceElement::queryConditions(
	tstring& aSourcePath,
	std::vector<tstring>& aConditionNameList)
{
	return E_NOTIMPL;
}   //  end queryConditions


//-----------------------------------------------------------------------------
//	AeAddressSpaceRoot constructor
//
AeAddressSpaceRoot::AeAddressSpaceRoot(EnumAddressSpaceType anAddressSpaceType):
	AddressSpaceRoot(anAddressSpaceType,
					 new AeAddressSpaceElement(tstring(_T("AeRoot")), 0, 0, 0))
{
}   //  end constructor


//-----------------------------------------------------------------------------
//	AeAddressSpaceRoot constructor
//
AeAddressSpaceRoot::AeAddressSpaceRoot(void):
	AddressSpaceRoot(EnumAddressSpaceType_OBJECT_STRING_BASED,
					 new AeAddressSpaceElement(tstring(_T("AeRoot")),
											   0, 0, 0))
{
}   //  end constructor


//-----------------------------------------------------------------------------
//	AeAddressSpaceRoot destructor
//
AeAddressSpaceRoot::~AeAddressSpaceRoot(void)
{
	m_elementsJanitor.lock();
	if (!m_elements.empty())
	{
		Application::Instance()->trace(EnumTraceLevel_INF, EnumTraceGroup_OPCSERVER, _T("~AeAddressSpaceRoot"),
									   _T("elements vector not empty"));
	}   //  end if
	m_elementsJanitor.unlock();
}   //  destructor


//-----------------------------------------------------------------------------
// Returns the string based adderss space element
//
long AeAddressSpaceRoot::queryAddressSpaceElementData(
	const tstring& anElementId,
	AddressSpaceElement*& anElement)
{
	anElement = NULL;
	return E_NOTIMPL;
}   //  end queryAddressSpaceElementData


//-----------------------------------------------------------------------------
// Returns the string based adderss space element (extended)
//
long AeAddressSpaceRoot::queryAddressSpaceElementDataEx(
	const tstring& anElementId,
	EnumAddressSpaceElementData& elementDataFilter,
	AddressSpaceElement*& anElement)
{
	anElement = NULL;
	return E_NOTIMPL;
}   //  end queryAddressSpaceElementDataEx


//-----------------------------------------------------------------------------
//	queryAddressSpaceElementChildren
//
//		-   Returns the list of children the of the string based addess space
//			element. returns E_NOTIMPL - Must be overridden to support string
//			based addess space
//
long AeAddressSpaceRoot::queryAddressSpaceElementChildren(
	const tstring& anElementId,
	std::vector<AddressSpaceElement*>& aChildrenList)
{
	aChildrenList.clear();
	return E_NOTIMPL;
}   //  end queryAddressSpaceElementChildren


//-----------------------------------------------------------------------------
// QueryConditions
//
long AeAddressSpaceRoot::queryConditions(
	unsigned long anElementUserData,
	tstring& aSourcePath,
	std::vector<tstring>& aConditionNames)
{
	m_elementsJanitor.lock();
	AeAddressSpaceElement* element = (AeAddressSpaceElement*)getElementFromArray(anElementUserData);

	if (element == NULL)
	{
		aConditionNames.clear();
		m_elementsJanitor.unlock();
		return S_FALSE;
	}   //  end if

	long result = element->queryConditions(aSourcePath, aConditionNames);
	m_elementsJanitor.unlock();
	return result;
}   //  end queryConditions

File: development/NET/src/server/Ae/ServerAeAddressSpaceElement.cs
using System;
using System.Collections;
using System.Runtime.InteropServices;
using Softing.OPCToolbox.OTB;
using Softing.OPCToolbox.Server;

namespace Softing.OPCToolbox.Server
{
	/// <summary>
	/// AeAddressSpaceElement is the base class for defining the OPC server's
	/// address space element.
	/// </summary>
	/// <include
	///		file='TBNS.doc.xml'
	///		path='//class[@name="AeAddressSpaceElement"]/doc/*'
	///	/>
	public class AeAddressSpaceElement : AddressSpaceElement
	{
		#region //	Public Constructors
		//--------------------------------

		/// <summary>
		/// Internal constructor for the AeAddressSpace Element
		/// </summary>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceElement"]/
		///			ctor[@name="AeAddressSpaceElement" or @name="AeAddressSpaceElement.default"]/doc/*'
		///	/>
		public AeAddressSpaceElement()
		{
		} //	end ctr


		/// <summary>
		/// Constructor for the AeAddressSpaceElement. For dynamically usage only
		/// </summary>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceElement"]/
		///			ctor[@name="AeAddressSpaceElement" or @name="AeAddressSpaceElement.all"]/doc/*'
		///	/>
		public AeAddressSpaceElement(
			string aName,
			uint anUserData,
			uint anObjectHandle,
			uint aParentHandle) :
				base(aName, anUserData, anObjectHandle, aParentHandle)
		{
		} //	end ctr


		//-
		#endregion

		#region //	Public Methods
		//------------------------

		/// <summary>
		///  Adds a namespace element as child to a namespace element
		/// </summary>
		/// <param name="aChild">Child to be added</param>
		/// <returns>
		/// true	- Child successfully added
		/// false	- Child was not added
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///		path='//class[@name="AeAddressSpaceElement"]/
		///		method[@name="AddChild"]/doc/*'
		///	/>
		public override bool AddChild(AddressSpaceElement aChild)
		{
			if (!HasChildren)
			{
				Application.Instance.Trace(
					EnumTraceLevel.WRN, EnumTraceGroup.OPCSERVER,
					"AeAddressSpaceElement.AddChild", "this element does not allow childred");
				return false;
			} //	end if

			AeAddressSpaceElement child = aChild as AeAddressSpaceElement;
			if (child == null)
			{
				Application.Instance.Trace(
					EnumTraceLevel.WRN, EnumTraceGroup.OPCSERVER,
					"AeAddressSpaceElement.AddChild", "Invalid child provided");
				return false;
			} //	end if

			//	only add the child if no handle assigned
			if (child.ObjectHandle == 0)
			{
				//	Prepare for the add addressSpace element call
				OTSAddressSpaceElementData data = new OTSAddressSpaceElementData();

				data.m_name = Marshal.StringToCoTaskMemUni(child.Name);
				data.m_itemID = Marshal.StringToCoTaskMemUni(
					Name + Application.Instance.AddressSpaceDelimiter + child.Name);

				data.m_hasChildren = Convert.ToByte(child.HasChildren);
				data.m_isBrowsable = Convert.ToByte(child.IsBrowsable);
				data.m_elementType = (byte) EnumAddressSpaceElementType.AE;
				data.m_userData = child.UserData;

				Application.Instance.AeAddressSpaceRoot.AddElementToArray(child);

				int result = OTBFunctions.OTSAddAddressSpaceElement(m_objectHandle, data, out child.m_objectHandle);

				Marshal.FreeCoTaskMem(data.m_name);
				Marshal.FreeCoTaskMem(data.m_itemID);

				if (!ResultCode.SUCCEEDED(result))
				{
					result = Application.Instance.AeAddressSpaceRoot.RemoveElementFromArray(child);
					System.Diagnostics.Debug.Assert(ResultCode.SUCCEEDED(result),
					                                "RemoveElementFromArray has failed");
					return false;
				} //	end if
			}
			else
			{
				Application.Instance.AeAddressSpaceRoot.AddElementToArray(child);
			} //	end if ... else

			return base.AddChild(aChild);
		} //	end AddChild


		/// <summary>
		/// Removes Child and its children if there are some
		/// </summary>
		/// <param name="aChild">AddressSpaceElement to be removed</param>
		/// <returns>
		/// true	- Child removed
		/// false	- Child was not removed
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceElement"]/
		///			method[@name="RemoveChild"]/doc/*'
		///	/>
		public override bool RemoveChild(AddressSpaceElement aChild)
		{
			AeAddressSpaceElement child = aChild as AeAddressSpaceElement;

			if (child == null)
			{
				Application.Instance.Trace(EnumTraceLevel.WRN, EnumTraceGroup.OPCSERVER,
				                           "AeAddressSpaceElement.RemoveChild", "Invalid child provided");
				return false;
			} //	end if

			if (child.HasChildren)
			{
				//	Remove the clildren of the child
				ArrayList otherChildren = child.GetChildren();

				foreach (AddressSpaceElement element in otherChildren)
				{
					child.RemoveChild(element);
				} //	end for
			} //	end if

			int result = OTBFunctions.OTSRemoveAddressSpaceElement(child.ObjectHandle);
			if (!ResultCode.SUCCEEDED(result))
			{
				Application.Instance.Trace(EnumTraceLevel.WRN, EnumTraceGroup.OPCSERVER,
				                           "AeAddressSpaceElement.RemoveChild",
				                           "OTSRemoveAddressSpaceElement has failed with code" + result.ToString());
				return false;
			} //	end if

			result = Application.Instance.AeAddressSpaceRoot.RemoveElementFromArray(child);
			if (!ResultCode.SUCCEEDED(result))
			{
				Application.Instance.Trace(EnumTraceLevel.WRN, EnumTraceGroup.OPCSERVER,
				                           "AeAddressSpaceElement.RemoveChild",
				                           "OTSRemoveAddressSpaceElement has failed with code" + result.ToString());
				return false;
			} //	end if

			return base.RemoveChild(aChild);
		} //	end RemoveChild


		/// <summary>
		/// Returns the element`s parent
		/// </summary>
		/// <returns>
		/// The parent of this element.
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceElement"]/
		///			method[@name="GetParent"]/doc/*'
		///	/>
		public override AddressSpaceElement GetParent()
		{
			return Application.Instance.AeAddressSpaceRoot.GetParent(this.ObjectHandle);
		} //	end GetParent


		/// <summary>
		/// Returns a list of children. This list is empty in case of an error or if no children are there to be delivered
		/// </summary>
		/// <returns>
		/// List of this element's children
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceElement"]/
		///			method[@name="GetChildren"]/doc/*'
		///	/>
		public override ArrayList GetChildren()
		{
			ArrayList children = new ArrayList();
			IntPtr ptrData = IntPtr.Zero;
			int count = 0;

			int res = OTBFunctions.OTSGetChildren(m_objectHandle, (byte) EnumAddressSpaceElementType.AE, out count, out ptrData);

			if (ResultCode.SUCCEEDED(res))
			{
				if (count > 0)
				{
					OTObjectData TypeOfObjData = new OTObjectData();
					int size = Marshal.SizeOf(TypeOfObjData);

					lock (Application.Instance.AeAddressSpaceRoot.ElementSyncRoot)
					{
						for (int i = 0; i < count; i++)
						{
							IntPtr currentPtr = new IntPtr(ptrData.ToInt64() + size*i);
							OTObjectData myData = (OTObjectData) Marshal.PtrToStructure(currentPtr, typeof (OTObjectData));
							children.Add(
								Application.Instance.AeAddressSpaceRoot.GetElementFromArray(myData.m_userData) as AeAddressSpaceElement);
						} //	end for
					}
				} //	end if

				OTBFunctions.OTFreeMemory(ptrData);
			} //	end if

			return children;
		} //	end GetChildren


		/// <summary>
		/// Provides the available conditions for this address space element.
		/// This virtual function has no default implementation. The user should override it with the appropriate implementation
		/// </summary>
		/// <param name="aSourcePath">
		/// the AE address space source path
		/// </param>
		/// <param name="aConditionNames">
		/// out parameter; filed by the method with the list of strings
		/// </param>
		/// <returns>
		/// EnumResultCode.E_NOTIMPL - this must be implemented if
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceElement"]/
		///			method[@name="QueryConditions"]/doc/*'
		///	/>
		public virtual int QueryConditions(
			string aSourcePath,
			out string[] aConditionNames)
		{
			aConditionNames = null;
			return (int) EnumResultCode.E_NOTIMPL;
		} //	end QueryConditions


		//-
		#endregion
	} //	end class AeAddressSpaceElement

	/// <summary>
	/// Implementation of the OPC-AE Address Space Root node
	/// </summary>
	/// <include
	///		file='TBNS.doc.xml'
	///		path='//class[@name="AeAddressSpaceRoot"]/doc/*'
	///	/>
	public class AeAddressSpaceRoot : AddressSpaceRoot
	{
		#region //	Public Constructors
		//-----------------------------

		/// <summary>
		/// Creates a specified type fo address space space root node.
		/// </summary>
		/// <param name="anAddressSpaceType">object, string or both - based address space type
		/// </param>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceRoot"]/
		///			ctor[@name="AeAddressSpaceRoot" or @name="AeAddressSpaceRoot.type"]/doc/*'
		///	/>
		public AeAddressSpaceRoot(EnumAddressSpaceType anAddressSpaceType) :
			base(anAddressSpaceType,
			     new AeAddressSpaceElement("AeRoot", 0, 0, 0))
		{
		} //	end constructor


		/// <summary>
		/// Creates an object space root node.
		/// </summary>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceRoot"]/
		///			ctor[@name="AeAddressSpaceRoot" or @name="AeAddressSpaceRoot.default"]/doc/*'
		///	/>
		public AeAddressSpaceRoot() : this(EnumAddressSpaceType.OBJECT_STRING_BASED)
		{
		} //	end constructor


		//-
		#endregion

		#region	//	Public Methods
		//------------------------

		/// <summary>
		/// Queries for a string based address space element
		/// </summary>
		/// <param name="anElementId">Address space path of the element</param>
		/// <param name="anElement">Out parameter - the element</param>
		/// <returns>
		/// EnumResultCode.E_NOTIMPL - Must be overridden to support string based address space
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceRoot"]/
		///			method[@name="QueryAddressSpaceElementData"]/doc/*'
		///	/>
		public override int QueryAddressSpaceElementData(string anElementId, out AddressSpaceElement anElement)
		{
			anElement = null;
			return (int) EnumResultCode.E_NOTIMPL;
		} //	end QueryAddressSpaceElementData


		/// <summary>
		/// Queries for a string based address space element (extended version)
		/// </summary>
		/// <param name="anElementId">Address space path of the element</param>
		/// <param name="elementDataFilter">Type of the element data to be filled in</param>
		/// <param name="anElement">Out parameter - the element</param>
		/// <returns>
		/// EnumResultCode.E_NOTIMPL - Must be overridden to support string based address space
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceRoot"]/
		///			method[@name="QueryAddressSpaceElementDataEx"]/doc/*'
		///	/>
		public override int QueryAddressSpaceElementDataEx(string anElementId, ref EnumAddressSpaceElementData elementDataFilter, out AddressSpaceElement anElement)
		{
			anElement = null;
			return (int) EnumResultCode.E_NOTIMPL;
		} //	end QueryAddressSpaceElementDataEx


		/// <summary>
		/// Returns the list of children of the string based addess space element.
		/// </summary>
		/// <param name="anElementId">Address space path of the element</param>
		/// <param name="aChildrenList">List to be filled with the children elements</param>
		/// <returns>
		/// EnumResultCode.E_NOTIMPL - Must be overridden to support string based addess space
		/// </returns>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceRoot"]/
		///			method[@name="QueryAddressSpaceElementChildren"]/doc/*'
		///	/>
		public override int QueryAddressSpaceElementChildren(string anElementId, ArrayList aChildrenList)
		{
			aChildrenList.Clear();
			return (int) EnumResultCode.E_NOTIMPL;
		} //	end QueryAddressSpaceElementChildren


		/// <summary>
		/// retireves the conditions strings for a specified <see cref="AeAddressSpaceElement"/>
		/// </summary>
		/// <include
		///		file='TBNS.doc.xml'
		///			path='//class[@name="AeAddressSpaceRoot"]/
		///			method[@name="QueryConditions"]/doc/*'
		///	/>
		public virtual int QueryConditions(
			uint anUserData,
			string aSourcePath,
			out string[] aConditionNames)
		{
			AeAddressSpaceElement element = (AeAddressSpaceElement) GetElementFromArray(anUserData);

			if (element != null)
			{
				return element.QueryConditions(aSourcePath, out aConditionNames);
			}
			else
			{
				aConditionNames = null;
				return (int) EnumResultCode.S_FALSE;
			} //	end if ... else
		} //	end QueryConditions


		//-
		#endregion
	} //	end class AeAddressSpaceRoot
} //	end namespace Softing.OPCToolbox.Server